src/include/duckdb/storage/compression/chimp/chimp_scan.hpp

```diff
@@ -0,0 +1,300 @@
+		}
+		// There are 8 leading zero values packed in one block, the block could not be entire filled
+		flag_one_count = AlignValue<idx_t, 8>(flag_one_count);
+		D_ASSERT(flag_one_count == leading_zero_block_size);
+#endif
+		duckdb_chimp::LeadingZeroBuffer<false> leading_zero_buffer;
+		leading_zero_buffer.SetBuffer(packed_data);
+		for (idx_t i = 0; i < leading_zero_block_size; i++) {
+			leading_zeros[i] =
+			    duckdb_chimp::ChimpDecompressionConstants::LEADING_REPRESENTATION[leading_zero_buffer.Extract()];
+		}
+		max_leading_zeros_to_read = leading_zero_block_size;
+		leading_zero_index = 0;
+	}
+
+	idx_t CalculatePackedDataCount() const {
+		idx_t count = 0;
+		for (idx_t i = 0; i < max_flags_to_read; i++) {
+			count += flags[1 + i] == duckdb_chimp::TRAILING_EXCEEDS_THRESHOLD;
+		}
+		return count;
+	}
+
+	template <class CHIMP_TYPE>
+	void LoadPackedData(uint16_t *packed_data, idx_t packed_data_block_count) {
+		for (idx_t i = 0; i < packed_data_block_count; i++) {
+			duckdb_chimp::PackedDataUtils<CHIMP_TYPE>::Unpack(packed_data[i], unpacked_data_blocks[i]);
+		}
+		unpacked_index = 0;
+		max_packed_data_to_read = packed_data_block_count;
+	}
+
+public:
+	idx_t index;
+	uint8_t flags[ChimpPrimitives::CHIMP_SEQUENCE_SIZE + 1];
+	uint8_t leading_zeros[ChimpPrimitives::CHIMP_SEQUENCE_SIZE + 1];
+	uint32_t leading_zero_index;
+	uint32_t unpacked_index;
+	duckdb_chimp::UnpackedData unpacked_data_blocks[ChimpPrimitives::CHIMP_SEQUENCE_SIZE];
+	idx_t max_leading_zeros_to_read;
+
+private:
+	idx_t max_flags_to_read;
+	idx_t max_packed_data_to_read;
+};
+
+template <class T>
+struct ChimpScanState : public SegmentScanState {
+public:
+	using CHIMP_TYPE = typename ChimpType<T>::type;
+
+	explicit ChimpScanState(ColumnSegment &segment) : segment(segment) {
+		auto &buffer_manager = BufferManager::GetBufferManager(segment.db);
+
+		handle = buffer_manager.Pin(segment.block);
+		auto dataptr = handle.Ptr();
+		// ScanStates never exceed the boundaries of a Segment,
+		// but are not guaranteed to start at the beginning of the Block
+		auto start_of_data_segment = dataptr + segment.GetBlockOffset() + ChimpPrimitives::HEADER_SIZE;
+		chimp_state.input.SetStream(start_of_data_segment);
+		auto metadata_offset = Load<uint32_t>(dataptr + segment.GetBlockOffset());
+		metadata_ptr = dataptr + segment.GetBlockOffset() + metadata_offset;
+		LoadGroup();
+	}
+
+	duckdb_chimp::Chimp128DecompressionState<CHIMP_TYPE> chimp_state;
+	BufferHandle handle;
+	data_ptr_t metadata_ptr;
+	idx_t total_value_count = 0;
+	ChimpGroupState group_state;
+
+	ColumnSegment &segment;
+
+	idx_t LeftInGroup() const {
+		return ChimpPrimitives::CHIMP_SEQUENCE_SIZE - (total_value_count % ChimpPrimitives::CHIMP_SEQUENCE_SIZE);
+	}
+
+	bool GroupFinished() const {
+		return (total_value_count % ChimpPrimitives::CHIMP_SEQUENCE_SIZE) == 0;
+	}
+
+	// Scan a group from the start
+	template <class CHIMP_TYPE>
+	void ScanGroup(CHIMP_TYPE *values, idx_t group_size) {
```

Comment:
I feel like ScanGroup and scanGroupPartial, could be written a bit more DRY. Maybe use a template parameter `bool WHOLE_GROUP`?